Display an address's primary name if available

alexgaribay opened the pull request:
![image](https://user-images.githubusercontent.com/911605/44688446-e6547880-aa19-11e8-96dc-b248512a93e9.png)

![image](https://user-images.githubusercontent.com/911605/44689015-ad1d0800-aa1b-11e8-8de3-c82b921977df.png)
(Not the real foundation wallet)

## Motivation

Some addresses have a name attached to them such as contract. This PR displays an address's primary name if available in the UI under the address's hash.

## Changelog

### Enhancements
* Add function to return an address's primary name
* Names are preloaded when fetching an address by hash

acravenho commented:
@alexgaribay The name should not be on its own line in my opinion. I'm open to suggestions but I feel we have a lot of horizontal space to play with and adding a new line increases the size of the overview box.

Maybe add this to the left of the transactions count.

@katibest @arthurra WDYT?

KronicDeth commented on apps/explorer/lib/explorer/chain.ex:
```diff
@@ -485,7 +485,7 @@ defmodule Explorer.Chain do
       from(
         address in Address,
-        preload: [:smart_contract, :contracts_creation_internal_transaction],
+        preload: [:smart_contract, :contracts_creation_internal_transaction, :names],
```
While you're modifying this line, could you sort the preloads?

KronicDeth commented:
@alexgaribay I fixed `gettext` and rebased on master for you.

katibest commented:
I think having it inline with the transactions would look nice! I'm not sure we need the box around it - I think maybe just a regular link color would be ok. We're very excited to have names!! 🎉  @alexgaribay  @acravenho

alexgaribay commented:
@katibest Are you saying keep it in the same place but make the styling the same as the text for transaction count?

alexgaribay commented:
![image](https://user-images.githubusercontent.com/911605/44857089-acab8980-ac34-11e8-9dfd-a31242ef13ae.png)
@katibest Is that what you were envisioning?

alexgaribay commented:
@acravenho @KronicDeth @feliperenan I've updated this branch to change how the text is display in the detail tile. I also went ahead and just updated it to display address names everywhere. Please take another look when you all can get a chance.

acravenho commented:
@alexgaribay WDYT about moving the name to the top line where it says "Contract Address Details"? This would be very similar to what we display for Tokens on the Token Details page. 

<img width="1129" alt="screen shot 2018-08-30 at 1 25 53 pm" src="https://user-images.githubusercontent.com/17620007/44868233-40db1800-ac58-11e8-9cb8-ea5760ecbf1c.png">

katibest commented:
@alexgaribay  Sorry I missed these! I would say inline with transactions. 

![44857089-acab8980-ac34-11e8-9dfd-a31242ef13ae](https://user-images.githubusercontent.com/20226535/44870223-9d8d0180-ac5d-11e8-89ce-74a0302db473.png)

Kind of like this. Let me know what ya think!

alexgaribay commented:
The PR has been updated to reflect the inline name.

acravenho commented on apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex:
```diff
@@ -28,6 +28,10 @@ defmodule BlockScoutWeb.TransactionView do
 
   def to_address_hash(%Transaction{to_address: %Address{hash: address_hash}}), do: address_hash
 
+  def to_address(%Transaction{to_address: nil, created_contract_address: %Address{} = address}), do: address
```
I'm receiving the following error when attempting to view a contract page: `no function clause matching in BlockScoutWeb.TransactionView.to_address/1`

alexgaribay replied:
Which page were you viewing?

acravenho replied:
```
15:12:09.284 [error] #PID<0.1130.0> running BlockScoutWeb.Endpoint terminated
Server: localhost:4000 (http)
Request: GET /address/0xb7cffe2ac19b9d5705a24cbe14fef5663af905a6/transactions
** (exit) an exception was raised:
    ** (FunctionClauseError) no function clause matching in BlockScoutWeb.TransactionView.to_address/1
```

I just checked again and it worked. 🤔

alexgaribay replied:
I tried to cover every spot but we were passing just address hashes around prior to the PR. Now we pass the data structs instead. This just requires preloading the data. If we see this issue pop up, it's because I missed a preload somewhere.

acravenho replied:
I'm unable to replicate the error. Sorry for the confusion.
